cpp/memilio/utils/multi_index_range.h

```diff
@@ -0,0 +1,162 @@
+/* 
+* Copyright (C) 2020-2023 German Aerospace Center (DLR-SC)
+*
+* Authors: René Schmieding
+*
+* Contact: Martin J. Kuehn <Martin.Kuehn@DLR.de>
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#ifndef MIO_UTILS_MULTI_INDEX_RANGE_H_
+#define MIO_UTILS_MULTI_INDEX_RANGE_H_
+
+#include "memilio/utils/index.h"
+
+namespace mio
+{
+
+/**
+ * @brief A Range that can be used to iterate over a MultiIndex.
+ * @tparam MultiIndex A type mio::Index<Categories...> for some set of categories.
+ */
+template <class MultiIndex>
+class MultiIndexRange
+{
+public:
+    /**
+     * @brief Construct a Range that can be used to iterate over all MultiIndices in the given dimensions.
+     * The range for each Index i in the MultiIndex is determined by [0, d_i), where d_i is the dimension of i.
+     * @param[in] dimensions A MultIndex that contains the dimension for each Category.
+     */
+    MultiIndexRange(const MultiIndex& dimensions)
+        : m_dimensions(dimensions)
+        , m_end([](const MultiIndex& dims) {
+            // set end to the first invalid index that is reached by increments of 1,
+            // i.e. 0 everywhere except for the most significant index (position 0),
+            // which is set to its dimension
+            auto end         = MultiIndex::Zero();
+            mio::get<0>(end) = mio::get<0>(dims);
+            return end;
+        }(dimensions))
+    {
+    }
+
+    /**
+     * @brief Iterator for MultiIndices.
+     */
+    class MultiIndexIterator
+    {
+    public:
+        using value_type = MultiIndex;
+        using reference  = const MultiIndex&;
```

Comment:
If we want a standart compliant [ForwardIterator](https://en.cppreference.com/w/cpp/named_req/ForwardIterator), we also need a default constructor, which i am not sure would make a lot of sense - what default value should dimensions have? All ones, as the minimum allowed dimension, or all zeroes, so we cannot increment without triggering an assert (which might be preferrable?).
Also, the type `reference` needs to be either `const value_type&`  or `const value_type&&`.

Reply:
Since Index consists of size_ts, we cannot really use it as difference. Should we use a flat index (e.g. signed long) as difference type? This may technically cause over/underflows when using many and/or large categories, but I assume we do not need to worry about that.

Reply:
all default constructed iterators are invalid. If you default construct a vector::iterator, you can't do anything with it except compare it to another default constructed vector::iterator, probably implemented as nullptr.

> the type reference needs to be either const value_type& or const value_type&&

true for LegacyForwardIterator, but there have always been iterator libraries that just ignore that and get as close to the concept as possible because it's much more useful, e.g. to write a transformiterator. Those iterators still work for usual implementations of (non-modifying) algorithms. The requirement has even been dropped for std::forward_iterator in C++ 20 (the default constructor as well if I read the concept correctly)

I shouldn't have said "fully compatible" but "usable" :) Without the alias types above the iterator may not be recognized as an iterator by some algorithms that check iterator_category for example. `pointer` and `difference_type` can just be `void` and `ptrdiff_t`, since they don't matter for forward iterators